src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm

```diff
@@ -151,6 +151,20 @@ - (nullable instancetype)initWithController:(MTRDeviceController *)controller
     return self;
 }
 
+- (void)fetchAttributeDataForAllDevices:(MTRDeviceControllerDataStoreClusterDataHandler)clusterDataHandler
+{
+    __block NSDictionary<NSNumber *, NSDictionary<MTRClusterPath *, MTRDeviceClusterData *> *> * clusterDataByNode = nil;
+    dispatch_sync(_storageDelegateQueue, ^{
+        if ([self->_storageDelegate respondsToSelector:@selector(valuesForController:securityLevel:sharingType:)]) {
+            NSDictionary<NSString *, id> * dataStoreSecureLocalValues = [self->_storageDelegate valuesForController:self->_controller securityLevel:MTRStorageSecurityLevelSecure sharingType:MTRStorageSharingTypeNotShared];
+
+            clusterDataByNode = [self _getClusterDataFromSecureLocalValues:dataStoreSecureLocalValues];
+        }
+    });
+
+    clusterDataHandler(clusterDataByNode);
```

Comment:
Two issues here:

* There's no reason to run the conversion procedure on the delegate queue.
* We're calling our callback even if we have no data, right?
  
```suggestion
    __block NSDictionary<NSString *, id> * dataStoreSecureLocalValues = nil;
    dispatch_sync(_storageDelegateQueue, ^{
        if ([self->_storageDelegate respondsToSelector:@selector(valuesForController:securityLevel:sharingType:)]) {
            dataStoreSecureLocalValues = [self->_storageDelegate valuesForController:self->_controller securityLevel:MTRStorageSecurityLevelSecure sharingType:MTRStorageSharingTypeNotShared];

            clusterDataByNode = [self _getClusterDataFromSecureLocalValues:dataStoreSecureLocalValues];
        }
    });

    if (dataStoreSecureLocalValues) {
        clusterDataHandler([self _getClusterDataFromSecureLocalValues:dataStoreSecureLocalValues]);
    }
```

Reply:
Done.